### PR TITLE
feat: Warn PostHog staff before continuing stale, costly conversations

### PR DESCRIPTION
**Pull request by webjunkie**
## Problem

When a conversation becomes large (100k+ tokens) and idle (1+ hour), its Anthropic prompt cache has likely expired. Continuing such a conversation re-processes the entire context at full input price instead of the ~10% cached rate, resulting in unexpected costs. PostHog staff should be warned before incurring this expense.

### Core logic (`@posthog/core`)
- Added `shouldWarnStaleCostlyConversation()` to decide when to warn: true when a conversation is both large (≥100k tokens by default) and stale (idle ≥60 min by default)
- Added `extractLastActivityAt()` to find the most recent event timestamp in a session, used as a heuristic for prompt-cache freshness
- Added `StaleCostlyThreshold` interface and `DEFAULT_STALE_COSTLY_THRESHOLD` constant documenting the pricing rationale (100k tokens = ~10% of 1M context window; 60 min = longer than Anthropic's max cache TTL to ensure cache is cold before warning)
- Added comprehensive unit tests covering threshold boundaries, null timestamps, clock skew, and default fallback

### UI layer (`@posthog/ui`)
- Added `useStaleConversationGate()` hook that composes the warning logic with view state:
  - Checks if user is PostHog staff
  - Tracks per-session acknowledgement (ephemeral, not persisted)
  - Manages dialog open/dismissed state
  - Disables the composer while the warning is active
- Added `StaleConversationCostDialog` component displaying the warning with context (tokens used, idle time, cumulative cost if available)
- Added `useStaleConversationGateStore` (Zustand) to track which sessions have acknowledged the warning
- Integrated into `SessionView`:
  - Dialog blocks composer input until dismissed or acknowledged
  - Dismissed state shows a reopen affordance
  - Tooltip explains why the composer is disabled
- Added unit tests for the store's immutable acknowledgement logic

## How did you test this?

- Added 7 parametrized unit tests for `shouldWarnStaleCostlyConversation` covering:
  - Large + stale → warn
  - Large + fresh → no warn
  - Small + stale → no warn
  - Boundary conditions (exactly at thresholds, one token/ms below)
  - Null timestamp (never warns)
  - Future timestamp / clock skew (treats as fresh)
  - Default threshold fallback
- Added 2 unit tests for `extractLastActivityAt` (empty list, most recent event)
- Added 3 unit tests for `useStaleConversationGateStore` (immutability, idempotence, isolation)
- All tests pass; no manual testing needed for this soft cost warning (non-blocking, staff-only, informational)

## Automatic notifications

- [ ] Publish to changelog?
- [ ] Alert Sales and Marketing teams?

https://claude.ai/code/session_01RkLQSsdj7o4wQjcmjM1SD1

### CODE REVIEW

**Review by pauldambra — APPROVED** [review]

File: packages/core/src/sessions/contextUsage.ts [hunk 1]
```diff
@@ -111,3 +111,69 @@ function extractBreakdown(msg: AcpMessage["message"]): ContextBreakdown | null {
...
+ * Best-effort "time of last activity" for a session: the emit timestamp of the
+ * most recent event, or null for an empty list. Heuristic proxy for
+ * prompt-cache freshness — `ts` is stamped on *any* AcpMessage (agent chunks,
+ * tool calls, client-side events), not only turns sent to the model, so a
+ * purely local event can reset it without the cache being refreshed. Good
+ * enough for a soft cost warning; not a billing signal.
+ */
+export function extractLastActivityAt(events: AcpMessage[]): number | null {
+  return events.length > 0 ? events[events.length - 1].ts : null;
+}
```
[the remark this answers is not in the document]
Comment (webjunkie) [reply packages/core/src/sessions/contextUsage.ts:186]: Fixed in ad61cb0 — `extractLastActivityAt` now scans for the maximum `ts` instead of taking the last element, so an out-of-order append can't misreport last activity. Used an explicit loop rather than `Math.max(...events.map(...))`: the spread form can overflow the stack on long event lists, which is exactly the large-conversation case this gate targets. Added an out-of-order test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/ui/src/features/sessions/useStaleConversationGate.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  const usedTokens = contextUsage?.used ?? 0;
+  const lastActivityAt = extractLastActivityAt(events);
+  const active =
+    isStaff &&
+    !acknowledged &&
+    shouldWarnStaleCostlyConversation({
+      usedTokens,
+      lastActivityAt,
+      now: Date.now(),
+    });
```
[the remark this answers is not in the document]
Comment (webjunkie) [reply packages/ui/src/features/sessions/useStaleConversationGate.ts:57]: Intentional for now. This is a soft, staff-only cost hint rather than a hard guard, and in practice returning to an idle tab reliably triggers a render here (window focus, the connectivity/streaming subscriptions, editor focus), so the gate re-evaluates on return. The worst case of the race is a single un-warned message on a conversation that was already going to pay the cold-cache cost — benign for a warning. A force-render timer that resets on every streamed event (since `lastActivityAt` changes each event) is more machinery than this warrants. Happy to add the timer if we observe the gate actually missing in practice.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/ui/src/features/sessions/components/StaleConversationCostDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+          This conversation holds about {formatTokensCompact(usedTokens)} tokens
+          and {activity}. Its prompt cache has likely expired, so your next
+          message re-processes the whole conversation at full input price
+          instead of the ~10% cached rate
+          {costUsd !== null ? ` (≈$${costUsd.toFixed(2)} spent so far)` : ""}.
+          Starting a new conversation avoids the cost — continue only if you
+          need this thread's context.
+        </AlertDialog.Description>
+
+        <Flex justify="end" gap="2" mt="4">
```
[the remark this answers is not in the document]
Comment (webjunkie) [reply packages/ui/src/features/sessions/components/StaleConversationCostDialog.tsx:47]: The gateway cost is USD-only today: the underlying field is `total_cost_usd` (surfaced as `$ai_total_cost_usd`), and `ContextUsage.cost.currency` is set to `"USD"` upstream in `claude-agent.ts` — so `$` is correct. If multi-currency billing ever lands, the cleaner fix is a shared currency formatter (`Intl.NumberFormat`) applied wherever cost is rendered, not threading `currency` into this one dialog — I'd rather do it there than half-wire it here. Leaving as-is for now.

---
_Generated by [Claude Code](https://claude.ai/code)_
